// redux-zero/svelte/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function shallowEqual(a, b) {
    for (var i in a)
        if (a[i] !== b[i])
            return false;
    for (var i in b)
        if (!(i in a))
            return false;
    return true;
}

function differs(a, b) {
    if (a && typeof a === "object") {
        return !shallowEqual(a, b);
    }
    return a !== b;
}
function getDiff(newData, oldData) {
    var diff = {};
    var changed = false;
    for (var key in newData) {
        if (differs(oldData[key], newData[key])) {
            changed = true;
            diff[key] = newData[key];
        }
    }
    return { diff: diff, changed: changed };
}

function bindActions(actions, store) {
    var bound = {};
    var _loop_1 = function (name_1) {
        bound[name_1] = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var ret = actions[name_1].apply(actions, [store.getState()].concat(args));
            if (ret != null)
                store.setState(ret);
        };
    };
    for (var name_1 in actions) {
        _loop_1(name_1);
    }
    return bound;
}

function getActions(store, actions) {
    return bindActions(typeof actions === "function" ? actions(store) : actions, store);
}
function connect(component, store, mapToProps) {
    update();
    component.on("destroy", store.subscribe(update));
    function update() {
        var _a = getDiff(mapToProps(store.getState()), component.get()), diff = _a.diff, changed = _a.changed;
        if (changed) {
            component.set(diff);
        }
    }
}

exports.connect = connect;
exports.getActions = getActions;
